### DISCUSSION TITLE
Realtime Channel as Component Prop

### DISCUSSION THREAD

**Discussion by harshcut** [supabase/supabase#13241] [Questions] [answered]
**Description**

I am building a real-time nextjs application with multiple components. I want to pass the supabase channel as prop to other components. When I use this `channel` prop in other components it logs 2x times in the component. 

[CodeSandbox Link](https://codesandbox.io/p/sandbox/channel-realtime-bkgtz4)

**`channel` initialized in parent component**

```tsx
const [channel, setChannel] = React.useState<RealtimeChannel>()

React.useEffect(() => {
  const channel = supabase.channel(id, { config: { broadcast: { ack: true } } }).subscribe()
  setChannel(channel)
  return () => {
    supabase.removeChannel(channel)
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
}, [])
```

**`channel` consumed in child component**

```tsx
export default function Child({ channel }: { channel: RealtimeChannel }) {
  React.useEffect(() => {
    channel.on('broadcast', { event: 'time' }, ({ payload }) => {
      console.log(payload) /* logs 2x times */
    })
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <button onClick={() => channel.send({ type: 'broadcast', event: 'time', payload: new Date() })}>
      send
    </button>
  )
}
```

**Comment by w3b6x9** [comment]
@harshcut I think it could be b/c you called `.on` even though the channel is already subscribed to in the parent component. Make sure that `.subscribe` is called last and no `.on`s are called after the fact.

**Reply by harshcut** [reply]
![image](https://user-images.githubusercontent.com/86551248/228908661-b97b4447-0652-4c5c-8859-464ddbf7c448.png)

**`channel` initialized in parent component**

```tsx
const [channel, setChannel] = React.useState<RealtimeChannel>();

React.useEffect(() => {
  const channel = supabase.channel("input-room-name", {
    config: { broadcast: { self: true } },
  });
  setChannel(channel);
  return () => {
    /* don't unsubscribe as channel is not subscribed yet */
  };
}, []);
```
**`channel` consumed in child component**

```tsx
export default function Child({ channel }: { channel: RealtimeChannel }) {
  React.useEffect(() => {
    channel.on('broadcast', { event: 'time' }, ({ payload }) => {
      console.log(payload)
    })
    channel.subscribe()
    return () => {
      channel.unsubscribe()
    }
  }, [])

  return (
    <button onClick={() => channel.send({ type: 'broadcast', event: 'time', payload: new Date() })}>
      send
    </button>
  )
}
```

**Comment by harshcut** [comment, marked as answer]
@w3b6x9 I have already tested few things regarding this issue. The problem is caused by the React 18's `useEffect` hook running two times i.e. mounting, unmounting, and mounting again.

It looks like the `channel.on` listener is added two times. There should be a way to remove the `on` listener individually. Turing off React's strict mode does work but isn't a full-proof solution.

**Comment by w3b6x9** [comment]
@harshcut take a look at this: https://github.com/supabase/realtime-js/issues/169#issuecomment-1223026406

**Reply by harshcut** [reply]
I have already tried using `removeChannel` and `removeAllChannels` methods. It does not work and gives the [same error](https://github.com/orgs/supabase/discussions/13241#discussioncomment-5480175) in browser console. Here is a [CodeSandbox Link](https://codesandbox.io/p/sandbox/channel-realtime-forked-ijtcln?file=%2Fsrc%2FList.tsx) to reproduce.

**Reply by harshcut** [reply]
@w3b6x9 any update on this?

**Reply by w3b6x9** [reply]
Why don't you just start another channel in the child component instead?

**Reply by harshcut** [reply]
I could definitely do that. I thought initializing `channel` in parent would make much sense rather than initializing it in multiple child components.
